server/app/Pool.hs

```diff
@@ -0,0 +1,94 @@
+{-# OPTIONS_GHC -fno-warn-incomplete-uni-patterns #-}
+
+module Pool
+( HandleSet (..),
+  Pool (..),
+  dumpStatus,
+  assignJob,
+  finishJob,
+  removeWorker,
+) where
+
+import Control.Concurrent.STM (STM, TVar, atomically, readTVar, retry, writeTVar)
+import qualified Data.Foldable as F
+import Data.IntMap (IntMap)
+import qualified Data.IntMap as IM
+import qualified Data.List as List
+import Message (Id)
+import System.IO (Handle, hFlush, hPrint, hPutStrLn, stdout)
+import System.Process (ProcessHandle, terminateProcess)
+
+data HandleSet = HandleSet
+  { handleProcess :: ProcessHandle,
+    handleArgIn :: Handle,
+    handleMsgOut :: Handle
+  }
+
+data Pool = Pool
+  { poolLimit :: Int,
+    poolNext :: Int,
+    poolStatus :: IntMap (Bool, Maybe Id),
+    poolHandles :: [(Int, HandleSet)]
+  }
+
+dumpStatus :: TVar Pool -> IO ()
+dumpStatus ref = do
+  pool <- atomically (readTVar ref)
+  hPutStrLn stdout $ "poolLimit = " ++ show (poolLimit pool)
+  mapM_ (hPrint stdout) $ IM.toAscList (poolStatus pool)
+  hFlush stdout
+
+getAssignableWorker :: IntMap (Bool, Maybe Id) -> Maybe Id -> Maybe (Int, (Bool, Maybe Id))
+getAssignableWorker workers mid' = List.find (isAssignable . snd) . IM.toAscList $ workers
```

Comment:
FWIW, `Foldable.find @IntMap` searches in ascending key order 🙂

Reply:
Thanks. Unfortunately, in this case, we need to get Key (= Int) as well, so I couldn't use find directly on the map. toAscList was simply to fuse that key as element.